extensions/azurePublish/src/node/deploy.ts

```diff
@@ -195,4 +208,158 @@ export class BotProjectDeploy {
       }
     }
   }
+
+  /**
+   * link the bot channel registration with azure web app service
+   * @param absSettings the abs settings
+   * @param hostname the hostname of webapp which bot service would be linked to
+   */
+  private async linkBotWithWebapp(absSettings: any, hostname: string) {
+    let subscriptionId = '', resourceGroupName = '', botName = '';
+    if (!absSettings.resourceId || !hostname) {
+      this.logger({
+        status: BotProjectDeployLoggerType.DEPLOY_INFO,
+        message: 'Abs settings incomplete, skip linking bot with webapp ...'
+      });
+      return;
+    }
+    try {
+      subscriptionId = absSettings.resourceId.match(/subscriptions\/([\w-]*)\//)[1];
+      resourceGroupName = absSettings.resourceId.match(/resourceGroups\/([^\/]*)/)[1];
+      botName = absSettings.resourceId.match(/botServices\/([^\/]*)/)[1];
+    } catch (error) {
+      this.logger({
+        status: BotProjectDeployLoggerType.DEPLOY_INFO,
+        message: 'Abs settings resourceId is incomplete, skip linking bot with webapp ...'
+      });
+      return;
+    }
+
+    this.logger({
+      status: BotProjectDeployLoggerType.DEPLOY_INFO,
+      message: 'Linking bot with webapp ...'
+    });
+
+    const creds = new TokenCredentials(this.accessToken);
+    const azureBotSerivce = new AzureBotService(creds, subscriptionId);
+
+    const botGetResult = await azureBotSerivce.bots.get(resourceGroupName, botName);
+    if (botGetResult?._response?.status >= 300) {
+      this.logger({
+        status: BotProjectDeployLoggerType.DEPLOY_ERROR,
+        message: botGetResult._response?.bodyAsText,
+      });
+      throw createCustomizeError(AzurePublishErrors.ABS_ERROR, botGetResult._response?.bodyAsText);
+    }
+
+    const bot = botGetResult._response.parsedBody;
+    bot.properties.endpoint = `https://${hostname}.azurewebsites.net/api/messages`;
+
+    const botUpdateResult = await azureBotSerivce.bots.update(resourceGroupName, botName, {
+      tags: {
+        webapp: hostname
+      },
+      properties: bot.properties
+    });
+
+    if (botUpdateResult?._response?.status >= 300) {
+      this.logger({
+        status: BotProjectDeployLoggerType.DEPLOY_ERROR,
+        message: botUpdateResult._response?.bodyAsText,
+      });
+      throw createCustomizeError(AzurePublishErrors.ABS_ERROR, botUpdateResult._response?.bodyAsText);
+    }
+  }
+
+  /**
+   * Bind key vault settings to webapp
+   * @param absSettings
+   */
+  private async BindKeyVault(absSettings: any, hostname: string) {
+    const subscriptionId = absSettings.subscriptionId;
```

Comment:
Is it possible that the keyvault hint are in difference subscriptionId or resource group? I think to keep this part clean, we should only use the keyvault hint to generate the keyvault url. in the keyvault hint, we did have subscription and resource group information. Let's use that.